wrappers/services.go

```diff
@@ -66,37 +68,69 @@ func generateSnapServiceFile(app *snap.AppInfo) ([]byte, error) {
 	return genServiceFile(app), nil
 }
 
+func stopUserServices(cli *client.Client, inter interacter, services ...string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeout.DefaultTimeout))
+	defer cancel()
+	failures, err := cli.ServicesStop(ctx, services)
+	for _, f := range failures {
+		inter.Notify(fmt.Sprintf("Could not stop service %q for uid %d: %s", f.Service, f.Uid, f.Error))
+	}
+	return err
+}
+
+func startUserServices(cli *client.Client, inter interacter, services ...string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeout.DefaultTimeout))
+	defer cancel()
+	startFailures, stopFailures, err := cli.ServicesStart(ctx, services)
+	for _, f := range startFailures {
+		inter.Notify(fmt.Sprintf("Could not start service %q for uid %d: %s", f.Service, f.Uid, f.Error))
+	}
+	for _, f := range stopFailures {
+		inter.Notify(fmt.Sprintf("While trying to stop previously started service %q for uid %d: %s", f.Service, f.Uid, f.Error))
+	}
+	return err
+}
+
 func stopService(sysd systemd.Systemd, app *snap.AppInfo, inter interacter) error {
 	serviceName := app.ServiceName()
 	tout := serviceStopTimeout(app)
 
-	stopErrors := []error{}
+	var extraServices []string
 	for _, socket := range app.Sockets {
-		if err := sysd.Stop(filepath.Base(socket.File()), tout); err != nil {
-			stopErrors = append(stopErrors, err)
-		}
+		extraServices = append(extraServices, filepath.Base(socket.File()))
 	}
-
 	if app.Timer != nil {
-		if err := sysd.Stop(filepath.Base(app.Timer.File()), tout); err != nil {
-			stopErrors = append(stopErrors, err)
-		}
+		extraServices = append(extraServices, filepath.Base(app.Timer.File()))
 	}
 
-	if err := sysd.Stop(serviceName, tout); err != nil {
-		if !systemd.IsTimeout(err) {
-			return err
+	switch app.DaemonScope {
+	case snap.SystemDaemon:
+		stopErrors := []error{}
+		for _, service := range extraServices {
+			if err := sysd.Stop(service, tout); err != nil {
+				stopErrors = append(stopErrors, err)
+			}
 		}
-		inter.Notify(fmt.Sprintf("%s refused to stop, killing.", serviceName))
-		// ignore errors for kill; nothing we'd do differently at this point
-		sysd.Kill(serviceName, "TERM", "")
-		time.Sleep(killWait)
-		sysd.Kill(serviceName, "KILL", "")
 
-	}
+		if err := sysd.Stop(serviceName, tout); err != nil {
+			if !systemd.IsTimeout(err) {
+				return err
+			}
+			inter.Notify(fmt.Sprintf("%s refused to stop, killing.", serviceName))
+			// ignore errors for kill; nothing we'd do differently at this point
+			sysd.Kill(serviceName, "TERM", "")
```

Comment:
we need this kind of Kill-based code implemented in the agent, or not?

Reply:
As mentioned above, I believe this code path is a hold over from ancient versions of systemd and likely never gets run these days.  From the systemd.kill(5) man page:

> If processes still remain after the main process of a unit has exited or the delay configured via the TimeoutStopSec= has passed, the termination request is repeated with the SIGKILL signal or the signal specified via FinalKillSignal= (unless this is disabled via the SendSIGKILL= option). See kill(2) for more information.

and:

> SendSIGKILL=
>   Specifies whether to send SIGKILL (or the signal specified by FinalKillSignal=) to remaining processes after a timeout, if the normal shutdown procedure left processes of the service around.

Snapd provides no way for a snap to set `FinalKillSignal` or `SendSIGKILL` in their unit files, so there should be no opportunity for processes to remain after the stop operation completes.